src/garage/torch/algos/vpg.py

```diff
@@ -128,92 +134,149 @@ def train_once(self, itr, paths):
+            kl_after = self._compute_kl_constraint(obs)
             policy_entropy = self._compute_policy_entropy(obs)
 
-        average_returns = log_performance(itr,
-                                          TrajectoryBatch.from_trajectory_list(
-                                              self.env_spec, paths),
-                                          discount=self.discount)
-
         with tabular.prefix(self.policy.name):
-            tabular.record('LossBefore', loss.item())
-            tabular.record('LossAfter', loss_after.item())
-            tabular.record('dLoss', loss.item() - loss_after.item())
-            tabular.record('KLBefore', kl_before.item())
-            tabular.record('KL', kl.item())
-            tabular.record('Entropy', policy_entropy.mean().item())
+            tabular.record('/LossBefore', loss_before.item())
+            tabular.record('/LossAfter', loss_after.item())
+            tabular.record('/dLoss', loss_before.item() - loss_after.item())
+            tabular.record('/KLBefore', kl_before.item())
+            tabular.record('/KL', kl_after.item())
+            tabular.record('/Entropy', policy_entropy.mean().item())
 
-        self.baseline.fit(paths)
-        return np.mean(average_returns)
+        self._old_policy.load_state_dict(self.policy.state_dict())
+
+        undiscounted_returns = log_performance(
+            itr,
+            TrajectoryBatch.from_trajectory_list(self.env_spec, paths),
+            discount=self.discount)
+        return np.mean(undiscounted_returns)
+
+    def _train(self, obs, actions, rewards, advantages):
+        """Train the algorithm with minibatch.
+
+        Args:
+            obs (torch.Tensor): Observation from the environment.
+            actions (torch.Tensor): Actions fed to the environment.
+            rewards (torch.Tensor): Acquired rewards.
+            advantages (torch.Tensor): Advantage value at each step.
+
+        Returns:
+            torch.Tensor: Calculated mean scalar value of loss
+
+        """
+        loss = self._compute_loss_with_adv(obs, actions, rewards, advantages)
+
+        self._optimizer.zero_grad()
+        loss.backward()
+
+        self._optimize(obs, actions, rewards, advantages)
 
-    def _compute_loss(self, itr, obs, actions, rewards, valids, baselines):
+        return loss
+
+    def _compute_loss(self, obs, actions, rewards, valids, baselines):
         """Compute mean value of loss.
 
         Args:
-            itr (int): Iteration number.
             obs (torch.Tensor): Observation from the environment.
-            actions (torch.Tensor): Predicted action.
-            rewards (torch.Tensor): Feedback from the environment.
-            valids (list[int]): Array of length of the valid values.
+            actions (torch.Tensor): Actions fed to the environment.
+            rewards (torch.Tensor): Acquired rewards.
+            valids (list[int]): Numbers of valid steps in each paths.
```

Comment:
why is valids a list and not a tensor

Reply:
Because `valids` is just length of the valid steps, so I think np.array is enough to save the values. Also, `valids` is not used as a matrix or vectors, It is just used in for loop or indexing